WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/ShippingLabelFormViewController.swift

```diff
@@ -110,16 +118,37 @@ private extension ShippingLabelFormViewController {
-                                                                           type: .origin,
-                                                                           address: shippingLabelAddress,
-                                                                           completion: { [weak self] (shippingLabelAddress) in
-                                                                            guard let self = self else { return }
-                                                                            self.viewModel.handleOriginAddressValueChanges(address: shippingLabelAddress)
-                                                                           })
-            self.navigationController?.pushViewController(shippingAddressVC, animated: true)
+            self.viewModel.validateAddress(type: .origin) { [weak self] (validationState, response) in
+                guard let self = self else { return }
+                let shippingLabelAddress = self.viewModel.originAddress
+                switch validationState {
+                case .validated:
+                    self.viewModel.handleOriginAddressValueChanges(address: shippingLabelAddress,
+                                                                   validated: true)
+                case .suggestedAddress:
+                    let vc = ShippingLabelSuggestedAddressViewController(type: .origin,
+                                                                         address: shippingLabelAddress,
+                                                                         suggestedAddress: response?.address) { [weak self] (newShippingLabelAddress) in
+                        self?.viewModel.handleOriginAddressValueChanges(address: newShippingLabelAddress,
+                                                                        validated: true)
+                    }
+                    self.navigationController?.pushViewController(vc, animated: true)
+                default:
+                    let shippingAddressVC = ShippingLabelAddressFormViewController(siteID: self.viewModel.siteID,
+                                                                                   type: .origin,
+                                                                                   address: shippingLabelAddress,
+                                                                                   completion: { [weak self] (newShippingLabelAddress) in
+                                                                                    guard let self = self else { return }
+                            self.viewModel.handleOriginAddressValueChanges(address: newShippingLabelAddress, validated: true)
+
+                                                                                   })
+                    self.navigationController?.pushViewController(shippingAddressVC, animated: true)
+                }
```

Comment:
nit: 2 levels of callback closures are hard to read. Maybe we can move code for VC inits and callbacks into separate methods? Like:
```swift
case .suggestedAddress:
    self.showSuggestedAddressVC(for: shippingLabelAddress, suggestedAddress: response?.address)
```

Reply:
Sounds good to me 👍 269345410d8a3d573635ad5ad6b7f7d87f04f581